sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -513,6 +513,16 @@ ssize_t coap_build_reply(coap_pkt_t *pkt, unsigned code,
         }
     }
 
+    uint32_t no_response;
+    if (coap_opt_get_uint(pkt, COAP_OPT_NO_RESPONSE, &no_response) == 0) {
+        uint8_t mask = 1 << (code >> 6);
+
+        /* option contains bitmap of disinterest */
+        if ((type == COAP_TYPE_NON) && (no_response & mask)) {
+            return 0;
+        }
```

Comment:
Why would we need to adjust the callers if we didn't need to adjust the callers for the 0 value?

```suggestion
        /* option contains bitmap of disinterest */
        if (no_response & mask) {
            if (coap_get_type(pkt) == COAP_TYPE_NON) {
                return 0;
            } else {
                /* There is an immediate ACK response, but it is an empty
                 * response */
                code = COAP_CODE_EMPTY;
                len -= tkl;
                tkl = 0;
                payload_len = 0;
            }
        }
```

I've only given this a bit of sloppy testing, but I think it'd do the right thing.

Reply:
[`coap_reply_simply()`](https://github.com/RIOT-OS/RIOT/blob/master/sys/net/application_layer/nanocoap/nanocoap.c#L470) will add the payload marker to the header regardless.

Reply:
Ah, reading `coap_build_reply`'s docs makes it a bit clearer how this works. (The term "payload" was misleading me.)

This can probably be mitigated by not comparing to 0 but to a criterion the nanocoap library provides, like `>= COAP_BUILD_REPLY_SUCCESSFUL` (which would be the constant size of the CoAP header). (There is a corner case where a "payload" length of 0 is requested, and just that is returned even in the case of an actual response, but it makes no practical difference as such an application has nowhere to write anyway). That design isn't portable to other serializations of CoAP, but AIU that's not a goal of nanocoap_sock anyway.

Reply:
But should that be part of this PR?

Reply:
Ah I think your solution will work: The caller may copy the payload into the response buffer, but it is in vain: the length returned by `coap_build_reply()` is still authoritative so we will only send the empty ACK.

Reply:
Good point -- it's just doing needless work (wow, gotta be a full hundred byte memcpy! -- but not doing memory corruption), so there's no urgent need to check.

Starting a final review...